src/java/org/apache/cassandra/index/sai/disk/v2/sortedbytes/TrieRangeIterator.java

```diff
@@ -0,0 +1,266 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.disk.v2.sortedbytes;
+
+
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+
+import com.google.common.collect.AbstractIterator;
+
+import org.apache.cassandra.index.sai.disk.v1.TrieTermsDictionaryReader;
+import org.apache.cassandra.io.tries.Walker;
+import org.apache.cassandra.io.util.Rebufferer;
+import org.apache.cassandra.utils.Pair;
+import org.apache.cassandra.utils.SizedInts;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+
+/**
+ * Thread-unsafe term -> long trie range iterator.
+ * The payload may be changed to something else.
+ */
+public class TrieRangeIterator<Concrete extends TrieRangeIterator<Concrete>> extends Walker<Concrete>
+{
+    protected final ByteSource limit;
+    protected IterationPosition stack;
+    protected long next;
+    final TrieTermsDictionaryReader.TransitionBytesCollector collector = new TrieTermsDictionaryReader.TransitionBytesCollector();
+    final boolean inclEnd;
+
+    public static class IterationPosition
+    {
+        public long node;
+        public int childIndex;
+        public int limit;
+        public IterationPosition prev;
+
+        public IterationPosition(long node, int childIndex, int limit, IterationPosition prev)
+        {
+            super();
+            this.node = node;
+            this.childIndex = childIndex;
+            this.limit = limit;
+            this.prev = prev;
+        }
+
+        @Override
+        public String toString()
+        {
+            return String.format("[Node %d, child %d, limit %d]", node, childIndex, limit);
+        }
+    }
+
+    public TrieRangeIterator(Rebufferer source,
```

Comment:
A general note about this class. Our usage of it in the package is much simplified from the general case:

- `start != null`
- `end == null`
- `admitPrefix = true`
- `inclEnd = true`
If this is always going to be the case then the code paths not using these cases can be removed. This would considerably reduce the size of this class.

Reply:
Good idea ask Branimir to do so.